python/dgl/mock_sparse/elementwise_op_diag.py

```diff
@@ -0,0 +1,228 @@
+
+    Examples
+    --------
+    >>> D1 = DiagMatrix(torch.arange(1, 4))
+    >>> D2 = DiagMatrix(torch.arange(10, 13))
+    >>> D1 -D2
+    DiagMatrix(val=tensor([-9, -9, -9]),
+    shape=(3, 3))
+    """
+    if isinstance(D1, DiagMatrix) and isinstance(D2, DiagMatrix):
+        assert (
+            D1.shape == D2.shape
+        ), "The shape of diagonal matrix D1 {} and" "D2 {} must match".format(
+            D1.shape, D2.shape
+        )
+        return DiagMatrix(D1.val - D2.val)
+    raise RuntimeError(
+        "Elementwise subtraction between {} and {} is not "
+        "supported.".format(type(D1), type(D2))
+    )
+
+
+def mul(
+    D1: Union[DiagMatrix, float], D2: Union[DiagMatrix, float]
+) -> DiagMatrix:
+    """Elementwise multiplication.
+
+     Parameters
+     ----------
+     D1 : DiagMatrix or scalar
+         Diagonal matrix or scalar value
+     D2 : DiagMatrix or scalar
+         Diagonal matrix or scalar value
+
+    Returns
+     -------
+     DiagMatrix
+         diagonal matrix
+
+     Examples
+     --------
+     >>> D1 = DiagMatrix(torch.arange(1, 4))
+     >>> D2 = DiagMatrix(torch.arange(10, 13))
+     DiagMatrix(val=tensor([10, 22, 36]),
+     shape=(3, 3))
+     >>> D1 * 2.5
+     DiagMatrix(val=tensor([2.5000, 5.0000, 7.5000]),
+     shape=(3, 3))
+     >>> 2 * D1
+     DiagMatrix(val=tensor([2, 4, 6]),
+     shape=(3, 3))
+    """
+    if isinstance(D1, DiagMatrix) and isinstance(D2, DiagMatrix):
+        assert (
+            D1.shape == D2.shape
+        ), "The shape of diagonal matrix D1 {} and" "D2 {} must match".format(
+            D1.shape, D2.shape
+        )
+        return DiagMatrix(D1.val * D2.val)
+    return DiagMatrix(D1.val * D2)
+
+
+def div(D1: DiagMatrix, D2: Union[DiagMatrix, float]) -> DiagMatrix:
+    """Elementwise division.
+
+    Parameters
+    ----------
+    D1 : DiagMatrix
+        Diagonal matrix
+    D2 : DiagMatrix or scalar
+        Diagonal matrix or scalar value
+
+    Returns
+    -------
+    DiagMatrix
+        diagonal matrix
+
+    Examples
+    --------
+    >>> D1 = DiagMatrix(torch.arange(1, 4))
+    >>> D2 = DiagMatrix(torch.arange(10, 13))
+    >>> D1 / D2
+    >>> D1/D2
+    DiagMatrix(val=tensor([0.1000, 0.1818, 0.2500]),
+    shape=(3, 3))
+    >>> D1/2.5
+    DiagMatrix(val=tensor([0.4000, 0.8000, 1.2000]),
+    shape=(3, 3))
+    """
+    if isinstance(D1, DiagMatrix) and isinstance(D2, DiagMatrix):
+        assert (
+            D1.shape == D2.shape
+        ), "The shape of diagonal matrix D1 {} and" "D2 {} must match".format(
+            D1.shape, D2.shape
+        )
+        return DiagMatrix(D1.val / D2.val)
+    return DiagMatrix(D1.val / D2)
+
+
+def rdiv(D1: float, D2: DiagMatrix):
+    """Elementwise division.
+
+    Parameters
+    ----------
+    D1 : scalar
+        scalar value
+    D2 : DiagMatrix
+        Diagonal matrix
+    """
+    raise RuntimeError(
+        "Elementwise div between diagonal and dense matrix is not supported."
```

Comment:
The type annotation says D1 is a float but the error message talks about dense matrix.